### PR TITLE
fix setup for sidekiq 6.1.0

### PR DESCRIPTION
**Pull request by hmaack**
https://github.com/mperham/sidekiq/compare/v6.0.7...v6.1.0 shows internal API changes. The setup needs to initialize the fetcher instead of passing the fetch class. 🤷‍♂️ 

https://github.com/mperham/sidekiq/issues/4614 

I tried to fix that the easy way.

### CODE REVIEW

**Comment by hmaack** [comment]
fixed the rubocop issue but the CI has to run again 🕵️‍♂️

**Comment by shlima** [comment]
Waiting for this a lot

**Comment by jobwat** [comment]
is there any way to retrigger the CI? a empty commit?

**Comment by schonert** [comment]
@hmaack awesome! 

Is there anything we can do to push this forward?

@jobwat it seems that it only runs once when the pull request is created. It would require a new PR.
https://github.com/sensortower/sidekiq-throttled/blob/master/.github/workflows/ci.yml#L7

**Comment by hmaack** [comment]
The throttler gem supports versions below sidekiq 6.1, this PR is still not compatible for versions < 6.1. So in this case I would recommend either to check the version of sidekiq in the `#setup!` method and decide how to pass / setup the `Fetcher` or try another approach like to have a special `#setup` method for sidekiq >= 6.1. I would like to have an opinion from a contributer / maintainer about this. What do you think @ixti ? ✌️

**Comment by gnomex** [comment]
Hey guys, I'm waiting for this merge, there is anything I can help you to merge as soon as possible?

currently using as 
```
git 'https://github.com/9feet/sidekiq-throttled', branch: 'fixup/fetcher-init-sidekiq-6.1.0' do
  gem "sidekiq-throttled"
end
```

File: lib/sidekiq/throttled/fetch.rb
Comment (ixti) [inline lib/sidekiq/throttled/fetch.rb]: This method must be on singleton class for sidekiq 5.x. Thus, probably it is worth to add a module that will be included in sidekiq 6.x and extended in sidekiq 5.x

File: lib/sidekiq/throttled.rb
Comment (ixti) [inline lib/sidekiq/throttled.rb]: Can you please add a conditional here to check if we need one way or the other?

**Review by ixti — CHANGES_REQUESTED** [review]
Looks good to me, but I think we should do some conditionals for older and newer version of sidekiq

File: lib/sidekiq/throttled/fetch.rb
Comment (hmaack) [reply lib/sidekiq/throttled/fetch.rb]: Done without the conditional check but implemented as a module.

File: lib/sidekiq/throttled.rb
Comment (hmaack) [reply lib/sidekiq/throttled.rb]: Done and tested 👍

File: lib/sidekiq/throttled/fetch.rb
Comment (ixti) [inline lib/sidekiq/throttled/fetch.rb]:
```suggestion
      module BulkRequeue
        # Requeues all given units as a single operation.
        #
        # @see http://www.rubydoc.info/github/redis/redis-rb/master/Redis#pipelined-instance_method
        # @param [Array<Fetch::UnitOfWork>] units
        # @return [void]
        def bulk_requeue(units, _options)
          return if units.empty?
          
          Sidekiq.logger.debug { "Re-queueing terminated jobs" }
          Sidekiq.redis { |conn| conn.pipelined { units.each(&:requeue) } }
          Sidekiq.logger.info("Pushed #{units.size} jobs back to Redis")
        rescue => e
          Sidekiq.logger.warn("Failed to requeue #{units.size} jobs: #{e}")
        end
      end
      
      # https://github.com/mperham/sidekiq/commit/fce05c9d4b4c0411c982078a4cf3a63f20f739bc
      if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
        include BulkRequeue
      else
        extend BulkRequeue
      end
```

File: spec/sidekiq/throttled_spec.rb
Comment (ixti) [inline spec/sidekiq/throttled_spec.rb]:
```suggestion
      if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
        expect(Sidekiq.options[:fetch]).to be_a(Sidekiq::Throttled::Fetch)
      else
        expect(Sidekiq.options[:fetch]).to be Sidekiq::Throttled::Fetch
      end
```

**Review by ixti — CHANGES_REQUESTED** [review]
LGTM, I'd like to simplify code a bit. And in specs, I'd better see `WorkingClassHero.perform_async(...)` with further queue check rather than `redis.rpush`. In order to avoid spec duplication in different versions I think we should add something like:

``` ruby
shared_examples "provides bulk requeue API" do |scope|
  ...
end

if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
  describe "#bulk_requeue" do
    include_examples "provides bulk requeue API", fetcher
  end
else
  describe ".bulk_requeue" do
    include_examples "provides bulk requeue API", fetcher.class
  end
end

**Comment by hmaack** [comment]
Hi @ixti , thx for your suggestions. I switched all conditional checks to `Gem::Version.new(Sidekiq::VERSION) < Gem::Version.new("6.1.0")` because I got confused after reviewing the different styles in the code. I hope thats OK for you. Check bd2b2e0 and 0307564 they were also wrong at some points. So I tried to use a uniform approach.

File: spec/sidekiq/throttled/fetch_spec.rb
Comment (ixti) [inline spec/sidekiq/throttled/fetch_spec.rb:57]: I think this can be simplified to:

```suggestion
      3.times { |i| WorkingClassHero.perform_async(i) }
```

But I'm totally fine either way.

File: lib/sidekiq/throttled.rb
Comment (ixti) [inline lib/sidekiq/throttled.rb:111]: Please rename the method to `#setup_strategy!`, use indented style for conditionals, and Reduce comment to just a link to the changeset that changed API :D

```suggestion
      # @return [void]
      def setup_strategy!
        require "sidekiq/throttled/fetch"

        # https://github.com/mperham/sidekiq/commit/fce05c9d4b4c0411c982078a4cf3a63f20f739bc
        Sidekiq.options[:fetch] =
          if Gem::Version.new(Sidekiq::VERSION) < Gem::Version.new("6.1.0")
            Sidekiq::Throttled::Fetch
          else
            Sidekiq::Throttled::Fetch.new(Sidekiq.options)
          end
      end
```

**Review by ixti — APPROVED** [review]
Thank you!

**Comment by ixti** [comment]
Squash-merged in 700507c

**Comment by ixti** [comment]
Released as [v0.13.0](https://github.com/sensortower/sidekiq-throttled/releases/tag/v0.13.0)

**Pull request closed without merging**
